internal/service/indexer/transfer.go

```diff
@@ -0,0 +1,272 @@
+package indexer
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"log"
+	"math/big"
+
+	"github.com/dov-id/cert-integrator-svc/contracts"
+	"github.com/dov-id/cert-integrator-svc/internal/data"
+	"github.com/dov-id/cert-integrator-svc/internal/data/postgres"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/iden3/go-merkletree-sql/v2"
+	"gitlab.com/distributed_lab/logan/v3/errors"
+)
+
+func (i *Indexer) handleIssuerTransferLog(eventLog types.Log, client *ethclient.Client) error {
+	i.log.WithField("address", eventLog.Address.Hex()).Debugf("start handling transfer event")
+
+	issuer, err := contracts.NewTokenContract(eventLog.Address, client)
+	if err != nil {
+		return errors.Wrap(err, "failed to create new issuer instance")
+	}
+
+	event, err := issuer.ParseTransfer(eventLog)
+	if err != nil {
+		return errors.Wrap(err, "failed to parse transfer event data")
+	}
+
+	contract, err := i.ContractsQ.FilterByAddresses(event.Raw.Address.Hex()).Get()
+	if err != nil {
+		return errors.Wrap(err, "failed to get contract from database")
+	}
+
+	if contract == nil {
+		contract, err = i.ContractsQ.Insert(data.Contract{
+			Name:    IssuerContract,
+			Address: event.Raw.Address.Hex(),
+			Type:    data.ISSUER,
+		})
+		if err != nil {
+			return errors.Wrap(err, "failed to save new contract")
+		}
+	}
+
+	treeStorage := postgres.NewPGDBStorage(i.cfg.DB(), contract.Id)
+
+	mTree, err := merkletree.NewMerkleTree(i.ctx, treeStorage, 100)
+	if err != nil {
+		return errors.Wrap(err, "failed to create merkle tree")
+	}
+
+	if event.From.Hex() == ZeroAddress {
+		err = i.handleMint(mTree, event, int64(event.Raw.BlockNumber))
+		if err != nil {
+			return errors.Wrap(err, "failed to handle mint event")
+		}
+		return nil
+	}
+
+	err = i.handleTransfer(mTree, event, int64(event.Raw.BlockNumber), treeStorage)
+	if err != nil {
+		return errors.Wrap(err, "failed to handle transfer event")
+	}
+
+	return nil
+}
+
+func (i *Indexer) handleTransfer(mTree *merkletree.MerkleTree, event *contracts.TokenContractTransfer, blockNumber int64, treeStorage *postgres.Storage) error {
+	receiver := event.To.Big()
+
+	_, leafValue, _, err := mTree.Get(i.ctx, receiver)
+	if err != nil {
+		return errors.Wrap(err, "failed to get leaf")
+	}
+
+	value := leafValue.Int64() - 1
+	if value < 1 {
+		err = i.completelyDeleteKey(mTree, event, treeStorage)
+		if err != nil {
+			return err
+		}
+	} else {
```

Comment:
You can move this logic into a separate function and eliminate the else statement. We practice not using an `else`